engine/src/witness/hub.rs

```diff
@@ -0,0 +1,413 @@
+				Some(EventWrapper::BalancesTransfer { to, amount, from })
+			},
+			(TransactionFeePaid::PALLET, TransactionFeePaid::EVENT) => {
+				let TransactionFeePaid { actual_fee, tip, .. } =
+					event_details.as_event::<TransactionFeePaid>().unwrap().unwrap();
+				Some(EventWrapper::TransactionFeePaid { actual_fee, tip })
+			},
+			(ExtrinsicSuccess::PALLET, ExtrinsicSuccess::EVENT) => {
+				let ExtrinsicSuccess { .. } =
+					event_details.as_event::<ExtrinsicSuccess>().unwrap().unwrap();
+				Some(EventWrapper::ExtrinsicSuccess)
+			},
+			(AssetsTransferred::PALLET, AssetsTransferred::EVENT) => {
+				let AssetsTransferred { asset_id, from, to, amount } =
+					event_details.as_event::<AssetsTransferred>().unwrap().unwrap();
+				Some(EventWrapper::AssetsTransfer { asset_id, to, amount, from })
+			},
+			_ => None,
+		}
+		.map(|event| (event_details.phase(), event)),
+		Err(err) => {
+			error!("Error while parsing event: {:?}", err);
+			None
+		},
+	}
+}
+
+pub async fn proxy_added_witnessing(
+	epoch: Vault<cf_chains::Assethub, PolkadotAccountId, ()>,
+	header: Header<PolkadotBlockNumber, PolkadotHash, Vec<(Phase, EventWrapper)>>,
+) -> (Vec<(Phase, EventWrapper)>, BTreeSet<u32>) {
+	let events = header.data;
+	let proxy_added_broadcasts = proxy_addeds(header.index, &events, &epoch.info.0);
+
+	(events, proxy_added_broadcasts)
+}
+
+#[allow(clippy::type_complexity)]
+pub async fn process_egress<ProcessCall, ProcessingFut>(
+	epoch: Vault<cf_chains::Assethub, PolkadotAccountId, ()>,
+	header: Header<
+		PolkadotBlockNumber,
+		PolkadotHash,
+		(
+			(Vec<(Phase, EventWrapper)>, BTreeSet<u32>),
+			Vec<(PolkadotSignature, PolkadotBlockNumber)>,
+		),
+	>,
+	process_call: ProcessCall,
+	hub_client: DotRetryRpcClient,
+) where
+	ProcessCall: Fn(state_chain_runtime::RuntimeCall, EpochIndex) -> ProcessingFut
+		+ Send
+		+ Sync
+		+ Clone
+		+ 'static,
+	ProcessingFut: Future<Output = ()> + Send + 'static,
+{
+	let ((events, mut extrinsic_indices), monitored_egress_data) = header.data;
+
+	let monitored_egress_ids = monitored_egress_data
+		.into_iter()
+		.map(|(signature, _)| signature)
+		.collect::<BTreeSet<_>>();
+
+	// To guarantee witnessing egress, we are interested in all extrinsics that were successful
+	extrinsic_indices.extend(extrinsic_success_indices(&events));
+
+	let extrinsics: Vec<Bytes> = hub_client.extrinsics(header.hash).await;
+
+	for (extrinsic_index, tx_fee) in transaction_fee_paids(&extrinsic_indices, &events) {
+		let xt = extrinsics.get(extrinsic_index as usize).expect(
+			"We know this exists since we got
+	this index from the event, from the block we are querying.",
+		);
+		let mut xt_bytes = xt.0.as_slice();
+
+		match AssethubUncheckedExtrinsic::decode(&mut xt_bytes) {
+			Ok(unchecked) =>
+				if let Some(signature) = unchecked.signature() {
+					if monitored_egress_ids.contains(&signature) {
+						tracing::info!(
+							"Witnessing transaction_succeeded. signature: {signature:?}"
```

Comment:
Is this differentiated between the dot log in anyway? Can't seem to see any spans /instrumenting that would add the context of which chain